### PR TITLE
Use T-digest percentile since regular percentile in Pinot is super expensive.

### PR DESCRIPTION
**Pull request by buchi-busireddy**
* This PR also makes the percentile function configurable for Pinot request handlers. 
* Using new metrics library and report Pinot query latency with handler tag.
* This also pushes the tenantColumnName config to the request handler level since that's where it should ideally belong.
* Another change handled is making the slowQueryThreshold configurable, per handler.

### CODE REVIEW

File: query-service-client/build.gradle.kts [hunk 1]
```diff
@@ -12,5 +12,5 @@ dependencies {
   // Logging
   implementation("org.slf4j:slf4j-api:1.7.30")
   // Config
-  implementation("com.typesafe:config:1.3.2")
```
Comment (tim-mwangi) [inline query-service-client/build.gradle.kts:15]: while you are can you upgrade `io.grpc` dependencies to `1.31.1` and grpc utils to `0.1.4` and then get rid of the snyk ignore message? Keep the .snyk file but remove the vulnerability info. See https://github.com/hypertrace/java-grpc-utils/pull/4

File: query-service/src/main/resources/configs/common/application.conf [hunk 2]
```diff
@@ -18,6 +17,7 @@ service.config = {
       type = pinot
       clientConfig = zookeeper
       requestHandlerInfo = {
+        tenantColumnName = tenant_id
```
Comment (tim-mwangi) [inline query-service/src/main/resources/configs/common/application.conf:20]: Why did you move this into the request handler info? Seems wrong to me since it opens up more config bug opportunities.

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java [hunk 3]
```diff
@@ -118,7 +149,7 @@ public void handleRequest(
       QueryContext queryContext,
       QueryRequest request,
       QueryResultCollector<Row> collector,
-      RequestAnalyzer requestAnalyzer) {
+      RequestAnalyzer requestAnalyzer) throws Exception {
```
Comment (tim-mwangi) [inline query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:152]: Who is throwing the exception and can the be wrapped into a runtime exception?

File: query-service-client/build.gradle.kts (hunk 1 above)
Comment (buchi-busireddy) [reply query-service-client/build.gradle.kts:15]: Done.

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java (hunk 3 above)
Comment (buchi-busireddy) [reply query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java]: Removed it now and cleaned up the code a bit.

File: query-service/src/main/resources/configs/common/application.conf (hunk 2 above)
Comment (buchi-busireddy) [reply query-service/src/main/resources/configs/common/application.conf:20]: Because this is the right design. 
Imagine there are different data stores are present and there are handlers for each store, table combo. Do you want to dictate the same column name across all tables and stores? The tenant_id column belongs at the handler level so that they can choose whatever they want to call it as.

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceImplConfig.java [hunk 4]
```diff
@@ -34,16 +28,12 @@ public void setQueryRequestHandlersConfig(List<Config> queryRequestHandlersConfi
...
-    this.tenantColumnName = tenantColumnName;
-  }
-
   public static class RequestHandlerConfig {
 
     private String name;
     private String type;
     private String clientConfig;
-    private Map<String, Object> requestHandlerInfo;
+    private Config requestHandlerInfo;
```
Comment (aaron-steinfeld) [inline query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceImplConfig.java:36]: Not that the old interface was any better, but would it make sense to make this a real interface rather than passing around an untyped map of config? Not sure where the assumptions about config keys are made, but moving that to as close to the file read as possible makes it less error prone and more debuggable.

File: query-service-impl/src/test/resources/application.conf [hunk 5]
```diff
@@ -42,6 +44,9 @@ service.config = {
       type = pinot
       clientConfig = zookeeper
       requestHandlerInfo = {
+        tenantColumnName = tenant_id
+        slowQueryThresholdMs = 100
```
Comment (tim-mwangi) [inline query-service-impl/src/test/resources/application.conf:48]: Why is this being configured to a request handler? Isn't a slow request a slow request regardless of where it's being executed?

File: query-service/src/main/resources/configs/common/application.conf (hunk 2 above)
Comment (tim-mwangi) [reply query-service/src/main/resources/configs/common/application.conf:20]: I'd understand there were multiple data sources but we are not there yet. Also if you wanna go this route you will need to make changes to helm-values repo.

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceImpl.java [hunk 6]
```diff
@@ -30,41 +28,25 @@ public QueryServiceImpl(QueryServiceImplConfig config) {
...
-                  config
-                      .getRequestHandlerInfo()
-                      .get(PinotBasedRequestHandler.VIEW_DEFINITION_CONFIG_KEY),
-              tenantColumnName));
-      RequestHandlerRegistry.get()
-          .register(
-              config.getName(),
-              new RequestHandlerInfo(
-                  config.getName(), PinotBasedRequestHandler.class, requestHandlerInfoConf));
+      boolean registered = RequestHandlerRegistry.get().register(config.getName(),
```
Comment (tim-mwangi) [inline query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceImpl.java:44]: This is hard to test?

File: query-service-impl/src/test/resources/application.conf (hunk 5 above)
Comment (buchi-busireddy) [reply query-service-impl/src/test/resources/application.conf:48]: Flexibility. Applying the same slow query threshold for all tables and data stores isn't ideal and would give lot of false positives. Since each view/table could be different and query patterns on them could be different, this would allow that flexibility.
There is a default anyways that's applicable for all and it's not mandatory to override. In fact, i'm not setting this at handler level for actual service. This is a test.

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceImpl.java (hunk 6 above)
Comment (buchi-busireddy) [reply query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceImpl.java:44]: Currently the way PinotClient is created and first connections are made, yes, it's not straight forward. Maybe in a next PR.

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceImplConfig.java (hunk 4 above)
Comment (buchi-busireddy) [reply query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceImplConfig.java:36]: The keys of this config are assumed in the implementations of `RequestHandler`.
The reason this is a generic untyped config is, each handler could look differently and might have different keys for configs. For example: A pinot RequestHandler vs Druid RequestHandler config could look totally different and might need different set of keys.
Having said that, there could be ways to improve this further. I'll think about this and we can fix in future PRs. Given that I haven't changed the conf files now, this shouldn't be a big regression.

Comment (aaron-steinfeld) [reply query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceImplConfig.java:36]: Agreed it's not a regression and that it needs to be supported per-handler type. The same place we translate to the appropriate handler, we could translate to the appropriate config. That could fairly easily happen inside the handler implementation too, I just prefer to separate concerns and the quicker a `Config` gets converted to a typed object, the easier it is to trace back issues. Nothing need be done right now, for sure.

File: query-service-impl/src/test/java/org/hypertrace/core/query/service/QueryServiceImplTest.java [hunk 7]
```diff
@@ -65,7 +65,6 @@ public void testBlankTenantColumnNameThrowsException() {
 
     // whitespace tenant id column name
     QueryServiceImplConfig queryServiceConfig2 = new QueryServiceImplConfig();
-    queryServiceConfig2.setTenantColumnName("   ");
```
Comment (tim-mwangi) [inline query-service-impl/src/test/java/org/hypertrace/core/query/service/QueryServiceImplTest.java:68]: These tests are not valid anymore, are they? Also where are the tests to check for the empty tenant id now.

File: query-service/src/main/resources/configs/common/application.conf (hunk 2 above)
Comment (buchi-busireddy) [reply query-service/src/main/resources/configs/common/application.conf:20]: Replied to you separately for this.

File: query-service-impl/src/test/java/org/hypertrace/core/query/service/QueryServiceImplTest.java (hunk 7 above)
Comment (buchi-busireddy) [reply query-service-impl/src/test/java/org/hypertrace/core/query/service/QueryServiceImplTest.java:68]: you're right. Removed. Thanks for pointing this.

Comment (buchi-busireddy) [reply query-service-impl/src/test/java/org/hypertrace/core/query/service/QueryServiceImplTest.java:68]: `org.hypertrace.core.query.service.pinot.PinotBasedRequestHandlerTest#testNullTenantIdQueryRequestContextThrowsNPE`

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java [hunk 8]
```diff
@@ -82,17 +93,40 @@ public String getName() {
...
+      this.percentileAggFunction = config.getString(PERCENTILE_AGGREGATION_FUNCTION_CONFIG);
+    }
+    LOG.info("Using {} function for percentile aggregations of handler: {}",
+        this.percentileAggFunction, name);
+
+    this.request2PinotSqlConverter =
+        new QueryRequestToPinotSQLConverter(viewDefinition, this.percentileAggFunction);
+
+    if (config.hasPath(SLOW_QUERY_THRESHOLD_MS_CONFIG)) {
+      this.slowQueryThreshold = config.getInt(SLOW_QUERY_THRESHOLD_MS_CONFIG);
```
Comment (inespo) [inline query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:118]: thanks! very nice

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java [hunk 9]
```diff
@@ -282,6 +285,8 @@ private String convertExpression2String(Expression expression, Params.Builder pa
         // COUNT(*) for ORDER BY
         if (functionName.equalsIgnoreCase("COUNT")) {
           return functionName + "(*)";
+        } else if (functionName.startsWith(PERCENTILE_PREFIX) && !PERCENTILE_PREFIX.equals(functionName)) {
```
Comment (inespo) [inline query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java:288]: I think abstracting the percentile at this level is correct. This allows us to potentially change storage without changing the QueryService API.
Should we allow both Percentile and ApproxPercentile though as a function? 
We can add it later as long as we can configure function -> implementation

File: query-service-impl/src/test/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandlerTest.java [hunk 10]
```diff
@@ -44,6 +50,80 @@ public void setUp() {
...
+                .build(),
+            Set.of(), Set.of("Trace.start_time_millis", "Trace.end_time_millis", "Trace.duration_millis",
+                "Trace.id", "Trace.tags"));
+        Assertions.assertTrue(cost.getCost() >= 0.0d && cost.getCost() < 1.0d);
+      }
+    }
+  }
+
+  @Test
+  public void testCanHandleNegativeCase() {
```
Comment (inespo) [inline query-service-impl/src/test/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandlerTest.java:102]: Nice!

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java (hunk 9 above)
Comment (buchi-busireddy) [reply query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java:288]: The value of config lets you add either Percentile or Approx. That's the config indeed :)
